lib/active_merchant/billing/gateways/ixopay.rb

```diff
@@ -67,55 +54,193 @@ def scrub(transcript)
+          xml.tag! 'firstName', split_names(address[:name])[0]
+          xml.tag! 'lastName',  split_names(address[:name])[1]
+        end
+
+        xml.tag! 'billingAddress1', address[:address1]
+        xml.tag! 'billingAddress2', address[:address2]
+        xml.tag! 'billingCity',     address[:city]
+        xml.tag! 'billingPostcode', address[:zip]
+        xml.tag! 'billingState',    address[:state]
+        xml.tag! 'billingCountry',  address[:country]
+        xml.tag! 'billingPhone',    address[:phone]
       end
 
-      def add_payment(post, payment)
+      def add_shipping_address(xml, address)
+        if address[:name]
+          xml.tag! 'shippingFirstName', split_names(address[:name])[0]
+          xml.tag! 'shippingLastName',  split_names(address[:name])[1]
+        end
+
+        xml.tag! 'shippingCompany',   address[:company]
+        xml.tag! 'shippingAddress1',  address[:address1]
+        xml.tag! 'shippingAddress2',  address[:address2]
+        xml.tag! 'shippingCity',      address[:city]
+        xml.tag! 'shippingPostcode',  address[:zip]
+        xml.tag! 'shippingState',     address[:state]
+        xml.tag! 'shippingCountry',   address[:country]
+        xml.tag! 'shippingPhone',     address[:phone]
       end
 
-      def parse(body)
-        {}
+      def new_transaction_id
+        SecureRandom.uuid
       end
 
-      def commit(action, parameters)
+      def commit(action, request, options={})
         url = (test? ? test_url : live_url)
-        response = parse(ssl_post(url, post_data(action, parameters)))
+
+        begin
+          raw_response = ssl_post(url, request, headers(request))
+        rescue StandardError => error
+          return response_from_request_error(action, error)
+        end
+
+        response = parse(action, raw_response)
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def response_from_request_error(action, error)
+        response = parse(action, error.response.body)
 
         Response.new(
           success_from(response),
           message_from(response),
           response,
           authorization: authorization_from(response),
-          avs_result: AVSResult.new(code: response["some_avs_response_key"]),
-          cvv_result: CVVResult.new(response["some_cvv_response_key"]),
           test: test?,
           error_code: error_code_from(response)
         )
       end
 
       def success_from(response)
+        response[:success] == 'true'
       end
 
       def message_from(response)
+        response[:message] || response[:return_type]
       end
 
       def authorization_from(response)
-      end
-
-      def post_data(action, parameters = {})
+        response[:reference_id] ? response[:reference_id] + '|' + response[:purchase_id] : nil
```

Comment:
Looking at [the Ixopay docs](https://gateway.ixopay.com/documentation/api#transaction-response), it seems that `reference_id` is always returned, so I'm not sure whether this conditional is needed.

Perhaps also consider updating this to use either string interpolation or `.join`.

Reply:
In general, I'd say it's a good idea to check for null instead of relying on the remote API, which is out of our control. If there were an unexpected response, which could always change in the future, the result would be a runtime error without the check. In this case I think a case where the reference ID is not present is the error case (we use ```authorization_from``` in that case as well).

Revised to use string interpolation.

Reply:
Neither the `reference_id` nor the `purchase_id` is a nested parameter, so I don't believe that an error would be raised if either was missing. It just wouldn't show up in the response. 

I tested this by removing the conditional from the `authorization_from` method:
```
def authorization_from(response)
  "#{response[:reference_id]}|#{response[:purchase_id]}"
end
```
I then altered the `successful_purchase_response` method in the unit test file and deleted the `reference_id` line. When I then ran the `test_successful_purchase` unit test, the `response.authorization` was `"|20191016-b2bef23a30b537b90fbe"`, which indicates that the `authorization_from` method doesn't need the conditional.

Reply:
I was actually referring to the original implementation; it had ```response[:reference_id] + '|'```, so there would be a runtime error if ```response[:reference_id]``` were nil since we would have tried to add nil to a string. With the latest version, the intent is to return ```nil``` in the case of an error (signaled by the reference ID not being present) rather than a string containing nothing except a vertical pipe.